modal/_utils/async_utils.py

```diff
@@ -544,3 +545,116 @@ async def producer(producer_id: int, iterable: Union[AsyncIterable[T], Iterable[
 
 async def callable_to_agen(awaitable: Callable[[], Awaitable[T]]) -> AsyncGenerator[T, None]:
     yield await awaitable()
+
+
+async def async_map(
+    input: Union[AsyncIterable[T], Iterable[T]],
+    async_mapper_func: Callable[[T], Awaitable[V]],
+    concurrency: int,
+) -> AsyncGenerator[V, None]:
+    input_queue: asyncio.Queue[Tuple[str, Union[T, None]]] = asyncio.Queue()
+    output_queue: asyncio.Queue[Tuple[str, Union[V, Exception]]] = asyncio.Queue()
+    output_event = asyncio.Event()
+
+    async def producer():
+        async for item in sync_or_async_iter(input):
+            await input_queue.put(("value", item))
+        await input_queue.put(("stop", None))
+
+    async def worker():
+        while True:
+            try:
+                event_type, item = await input_queue.get()
+                if event_type == "stop":
+                    break
+                if event_type == "value":
+                    if asyncio.iscoroutinefunction(async_mapper_func):
```

Comment:
Checking a function with iscoroutinefunction is not 100% water tight, as the function could be regular function that *returns* a coroutine, without being declared as `async def`, in which case this would return False.

A better way to do this is to actually call the function (without await) and check if the returned value with `inspect.isawaitable` and only then await it.